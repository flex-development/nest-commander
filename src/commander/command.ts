/**
 * @file commander - Command
 * @module nest-commander/commander/Command
 */

import {
  cast,
  define,
  fallback,
  get,
  isEmptyString,
  isFalsy,
  isUndefined,
  trim,
  unique,
  type IfUndefined,
  type Nullable,
  type Optional
} from '@flex-development/tutils'
import { Command as CommandBase, type CommandOptions } from 'commander'
import Example from './example'

/**
 * CLI command model.
 *
 * @see {@linkcode CommandBase}
 *
 * @extends {CommandBase}
 */
class Command extends CommandBase {
  /**
   * Attached subcommands.
   *
   * @public
   * @override
   * @readonly
   * @member {Command[]} commands
   */
  public override readonly commands: Command[]

  /**
   * Command examples.
   *
   * @public
   * @override
   * @readonly
   * @member {Example[]} examples
   */
  public readonly examples: Example[]

  /**
   * Parent command.
   *
   * @public
   * @override
   * @readonly
   * @member {Nullable<Command>} parent
   */
  public override parent: Nullable<Command>

  /**
   * Creates a new command instance.
   *
   * @param {string?} [cmd] - Command name
   */
  constructor(cmd?: string) {
    super(cmd)
    this.commands = []
    this.examples = []
    this.parent = null
  }

  /**
   * Add a prepared subcommand.
   *
   * See {@linkcode command} for creating an attached subcommand that inherits
   * settings from its parent.
   *
   * @see {@linkcode CommandBase}
   * @see {@linkcode CommandOptions}
   *
   * @public
   * @override
   *
   * @param {Command} cmd - Command instance
   * @param {CommandOptions?} [opts] - Command options
   * @param {boolean?} [opts.hidden] - Remove command from help text
   * @param {boolean?} [opts.isDefault] - Mark command as root command
   * @return {this} `this` command for chaining
   */
  public override addCommand(cmd: Command, opts?: CommandOptions): this {
    return super.addCommand(cmd, opts)
  }

  /**
   * Create a new unattached command.
   *
   * @public
   * @override
   *
   * @param {string?} [name] - Command name
   * @return {Command} New command
   */
  public override createCommand(name?: string): Command {
    return new Command(name)
  }

  /**
   * Add a new example.
   *
   * @public
   *
   * @param {string?} [text=''] - Example text
   * @param {string?} [prefix=''] - Example text prefix
   * @return {this} `this` command for chaining
   */
  public example(text: string = '', prefix: string = ''): this {
    return define(this, 'examples', {
      value: unique([
        ...this.examples,
        new Example(
          !text.startsWith(this.alias()) && !text.startsWith(this.name())
            ? `${this.name()} ${text}`.replaceAll(/ +$/g, '')
            : text,
          prefix
        )
      ])
    })
  }

  /**
   * Get `this` command version, or modify the existing version configuration.
   *
   * Flags `--version` and `-v` are auto-registered if option `flags` are not
   * specified.
   *
   * @public
   * @override
   *
   * @template V - Command version
   *
   * @param {V?} [version] - Command version
   * @param {string?} [flags='-v, --version'] - Option flags
   * @param {string?} [description='print version number'] - Option description
   * @return {IfUndefined<V, string, this>} Current command version if `version`
   * is not passed, or `this` command for chaining
   */
  public override version<V extends Optional<string> = undefined>(
    version?: V,
    flags?: string,
    description?: string
  ): IfUndefined<V, string, this> {
    return cast(
      isUndefined(version)
        ? get(this, '_version')
        : isEmptyString((version = cast(trim(version))))
        ? this
        : super.version(
            version!,
            fallback(flags, '-v, --version', isFalsy),
            fallback(description, 'print version number', isFalsy)
          )
    )
  }
}

export default Command
