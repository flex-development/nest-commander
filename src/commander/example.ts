/**
 * @file commander - Example
 * @module nest-commander/commander/Example
 */

import { ifelse, trim, trimStart } from '@flex-development/tutils'

/**
 * Command example model.
 *
 * @class
 */
class Example {
  /**
   * Example {@linkcode text} prefix.
   *
   * @example
   *  'cross-env FORCE_COLOR=3'
   *
   * @default ''
   *
   * @public
   * @instance
   * @member {string} prefix
   */
  public prefix: string

  /**
   * Example text.
   *
   * @public
   * @instance
   * @member {string} text
   */
  public text: string

  /**
   * Creates a new command example.
   *
   * @param {string} text - Example text
   * @param {string?} [prefix=''] - Example text prefix
   */
  constructor(text: string, prefix: string = '') {
    this.prefix = trim(prefix)
    this.text = trimStart(text)
  }

  /**
   * Get a string representation of `this` example.
   *
   * @public
   *
   * @return {string} String representation of `this` example
   */
  public toString(): string {
    return ifelse(this.prefix, `${this.prefix} ${this.text}`, this.text)
  }
}

export default Example
