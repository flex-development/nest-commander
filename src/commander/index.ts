/**
 * @file Entry Point - commander
 * @module nest-commander/commander
 */

export {
  Argument,
  CommanderError,
  Help,
  InvalidArgumentError,
  Option,
  type AddHelpTextContext,
  type AddHelpTextPosition,
  type CommandOptions,
  type ErrorOptions,
  type HelpContext,
  type HookEvent,
  type OptionValueSource,
  type OptionValues,
  type OutputConfiguration,
  type ParseOptions,
  type ParseOptionsResult
} from 'commander'
export { default as Command } from './command'
export { default as Example } from './example'
