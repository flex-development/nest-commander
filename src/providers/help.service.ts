/**
 * @file Providers - HelpService
 * @module nest-commander/providers/HelpService
 */

import { Argument, Command, Help, Option } from '#src/commander'
import {
  cast,
  fallback,
  flat,
  fork,
  ifelse,
  includes,
  join,
  select,
  template,
  trim,
  trimStart,
  type Nullable
} from '@flex-development/tutils'
import { Injectable } from '@nestjs/common'

/**
 * Encapsulates logic for displaying help text.
 *
 * @see {@linkcode Help}
 *
 * @class
 * @extends {Help}
 */
@Injectable()
class HelpService extends Help {
  /**
   * Maximum output text width.
   *
   * Long lines are wrapped to fit.
   *
   * @default fallback(process.stdout.columns, 100)
   *
   * @public
   * @override
   * @instance
   * @member {number} helpWidth
   */
  public override helpWidth: number

  /**
   * Indent size in single-spaced characters.
   *
   * @public
   * @instance
   * @member {number} tabsize
   */
  public tabsize: number

  /**
   * Creates a new help service.
   */
  constructor() {
    super()
    this.helpWidth = fallback(process.stdout.columns, 110)
    this.showGlobalOptions = false
    this.sortOptions = true
    this.sortSubcommands = true
    this.tabsize = 2
  }

  /**
   * Format command arguments.
   *
   * @public
   *
   * @param {Command} cmd - Command instance
   * @param {Help?} [helper=this] - Command help instance
   * @return {string} Formatted command arguments
   */
  public formatArguments(cmd: Command, helper: Help = this): string {
    return this.section(
      'Arguments',
      this.formatList(this.visibleArguments(cmd), cmd, helper)
    )
  }

  /**
   * Format a command description.
   *
   * @public
   *
   * @param {Command} cmd - Command instance
   * @return {string} Formatted command description
   */
  public formatDescription(cmd: Command): string {
    return this.section(
      'Description',
      this.wrap(
        this.indent() + this.commandDescription(cmd),
        this.helpWidth,
        this.tabsize
      )
    )
  }

  /**
   * Format command examples.
   *
   * @public
   *
   * @param {Command} cmd - Command instance
   * @return {string} Formatted command examples
   */
  public formatExamples(cmd: Command): string {
    return this.section(
      'Examples',
      cmd.examples.reduce((acc, example) => {
        /**
         * Parent command names.
         *
         * @const {string[]}
         */
        const names: string[] = []

        /**
         * Current comand parent.
         *
         * @var {Nullable<Command>} parent
         */
        let parent: Nullable<Command> = cmd.parent

        // get parent command names
        while (parent) {
          names.unshift(parent.name())
          parent = parent.parent
        }

        return template('{acc}{newline}{wrapped}', {
          acc,
          newline: ifelse(acc, '\n', acc),
          wrapped: this.wrap(
            template('{indent}$ {example}', {
              example: trimStart(`${join(names, ' ')} ${example.toString()}`),
              indent: this.indent()
            }),
            this.helpWidth,
            this.tabsize
          )
        })
      }, '')
    )
  }

  /**
   * Format global options.
   *
   * @public
   *
   * @param {Command} cmd - Command instance
   * @param {Help?} [helper=this] - Command help instance
   * @return {string} Formatted global options
   */
  public formatGlobalOptions(cmd: Command, helper: Help = this): string {
    return this.section(
      'Global Options',
      this.formatList(this.visibleGlobalOptions(cmd), cmd, helper)
    )
  }

  /**
   * Format help text.
   *
   * @public
   * @override
   *
   * @param {Command} cmd - Command instance
   * @return {string} Formatted help text
   */
  public override formatHelp(cmd: Command): string {
    return template('{text}\n', {
      text: trim(
        join(
          select(
            [
              this.formatDescription(cmd),
              this.formatUsage(cmd),
              this.formatExamples(cmd),
              this.formatArguments(cmd),
              this.formatOptions(cmd),
              ifelse(this.showGlobalOptions, this.formatGlobalOptions(cmd), ''),
              this.formatSubcommands(cmd)
            ],
            section => !!section
          ),
          '\n\n'
        )
      )
    })
  }

  /**
   * Convert a list of arguments, options, or subcommands to a formatted list.
   *
   * @public
   *
   * @param {ReadonlyArray<Argument | Command | Option>} list - List to format
   * @param {Command} cmd - Command instance
   * @param {Help?} [helper=this] - Command help instance
   * @return {string} Formatted list
   */
  public formatList(
    list: readonly (Argument | Command | Option)[],
    cmd: Command,
    helper: Help = this
  ): string {
    return list.reduce((acc, item) => {
      return template('{acc}{newline}{wrapped}', {
        acc,
        newline: ifelse(acc, '\n', acc),
        wrapped: this.wrap(
          this.formatListItem(item, cmd, helper),
          this.helpWidth - this.tabsize,
          this.padWidth(cmd, helper) + this.tabsize
        )
      })
    }, '')
  }

  /**
   * Convert an argument, option, or subcommand to a formatted list item.
   *
   * @public
   *
   * @param {Argument | Command | Option} item - Item to format
   * @param {Command} cmd - Command instance
   * @param {Help?} [helper=this] - Command help instance
   * @return {string} Formatted list item
   */
  public formatListItem(
    item: Argument | Command | Option,
    cmd: Command,
    helper: Help = this
  ): string {
    /**
     * Argument, option, or subcommand description.
     *
     * @var {string} description
     */
    let description: string = ''

    /**
     * Argument, option, or subcommand term.
     *
     * @var {string} term
     */
    let term: string = ''

    // get argument description and term
    if (item instanceof Argument) {
      description = this.argumentDescription(item)
      term = this.argumentTerm(item)
    }

    // get subcommand description and term
    if (item instanceof Command) {
      description = this.subcommandDescription(item)
      term = this.subcommandTerm(item)
    }

    // get option description and term
    if (item instanceof Option) {
      description = this.optionDescription(item)
      term = this.optionTerm(item)
    }

    // pad term
    term = term.padEnd(this.padWidth(cmd, helper) + this.tabsize)
    term = term.padStart(term.length + this.tabsize)

    return template('{term}{description}', {
      description: this.wrap(
        description,
        this.helpWidth - term.length,
        term.length
      ),
      term
    })
  }

  /**
   * Format command options.
   *
   * @public
   *
   * @param {Command} cmd - Command instance
   * @param {Help?} [helper=this] - Command help instance
   * @return {string} Formatted command options
   */
  public formatOptions(cmd: Command, helper: Help = this): string {
    return this.section(
      'Options',
      this.formatList(this.visibleOptions(cmd), cmd, helper)
    )
  }

  /**
   * Format subcommands.
   *
   * @public
   *
   * @param {Command} cmd - Command instance
   * @param {Help?} [helper=this] - Command help instance
   * @return {string} Formatted subcommands
   */
  public formatSubcommands(cmd: Command, helper: Help = this): string {
    return this.section(
      'Commands',
      this.formatList(this.visibleCommands(cmd), cmd, helper)
    )
  }

  /**
   * Format command usage.
   *
   * @public
   *
   * @param {Command} cmd - Command instance
   * @return {string} Formatted command usage
   */
  public formatUsage(cmd: Command): string {
    return `Usage\n${this.wrap(
      template('{indent}$ {usage}', {
        indent: this.indent(),
        usage: this.commandUsage(cmd)
      }),
      this.helpWidth,
      this.tabsize
    )}`
  }

  /**
   * Create an indentation string.
   *
   * @public
   *
   * @param {number?} [size=this.indent_size] - Indent size
   * @return {string} Indentation string
   */
  public indent(size: number = this.tabsize): string {
    return ' '.repeat(size)
  }

  /**
   * Get the option term to show in the list of options.
   *
   * @public
   * @override
   *
   * @param {Option} option - Command option instance
   * @return {string} Option term to show in list of options
   */
  public override optionTerm(option: Option): string {
    return option.flags.padStart(option.flags.length + (option.short ? 0 : 4))
  }

  /**
   * Create a help text section.
   *
   * @public
   *
   * @param {string} title - Section title
   * @param {string?} [text=''] - Section content
   * @return {string} Formatted help text section
   */
  public section(title: string, text: string = ''): string {
    return ifelse(trim(text), template('{title}\n{text}', { text, title }), '')
  }

  /**
   * Get an array containing visible subcommands.
   *
   * If visible, the `help` command will **always** be the last subcommand.
   *
   * @public
   * @override
   *
   * @param {Command} cmd - Command instance
   * @return {Command[]} Visible subcommands array
   */
  public override visibleCommands(cmd: Command): Command[] {
    return flat(
      fork(
        cast<Command[]>(super.visibleCommands(cmd)),
        subcommand => subcommand.name() !== 'help'
      )
    )
  }

  /**
   * Get an array containing visible command options.
   *
   * If visible, the `--help` and `--version` flags will **always** be the last
   * two options.
   *
   * @public
   * @override
   *
   * @param {Command} cmd - Command instance
   * @return {Option[]} Visible command options array
   */
  public override visibleOptions(cmd: Command): Option[] {
    return flat(
      fork(
        super.visibleOptions(cmd),
        option => !includes(['help', 'version'], option.name())
      )
    )
  }
}

export default HelpService
