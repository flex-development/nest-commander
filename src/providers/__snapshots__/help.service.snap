// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`unit:providers/HelpService > #formatHelp > should return formatted help text 1`] = `
"Description
  https://github.com/felixge/node-dateformat

Usage
  $ dateformat [options] [command] [date]

Examples
  $ dateformat
  $ dateformat --mask=isoTime

Arguments
  date                date string or timestamp

Options
  -g, --gmt           convert local time to gmt time (default: false)
  -m, --mask <mask>   date format (default: \\"ddd mmm dd yyyy HH:MM:ss\\")
  -u, --utc           convert local time to utc time (default: false)
      --verbose       enable debug output
  -h, --help          display help for command
  -v, --version       print version number

Commands
  timezone|tz [date]  get proper timezone abbreviation or timezone offset
  help [command]      display help for command
"
`;

exports[`unit:providers/HelpService > #formatHelp > should return formatted help text for subcommand 1`] = `
"Description
  get proper timezone abbreviation or timezone offset

Usage
  $ dateformat timezone|tz [options] [date]

Examples
  $ dateformat timezone
  $ dateformat timezone 2023-08-21T21:45:48.850Z
  $ dateformat tz
  $ dateformat tz 2023-08-23T00:46:32.292Z

Arguments
  date        date string

Options
  -h, --help  display help for command
"
`;
